### ISSUE TITLE
Curated list of Good Design Principles w.r.t Code

### ISSUE THREAD

**Issue by AmitKumarDas** [CloudByteStorages/InfraSmart#5]
- <b> Rules based programming </b>

```
  - http://blog.acolyer.org/2016/01/19/dcft/
```
- <b>Wisdom from OfHeapArray</b>

```
  - provides a default instance via final static
  - provides lots of interfaces for specialized gets
  - defines specific exception
  - provides interface that defines contracts
  - implements the contracts in a default class
  - private constructor
  - completely modularized
  - No hard codings at all

  - reference: https://github.com/nerds-central/SonicFieldRepo/
```
- <b> Distributed Systems </b>

```
  - Read the Dynamo paper from Amazon
  - Peter Bailis' blogs
  - Morning Paper blog
```
- <b> Hypermedia APIs </b>

```
  - http://www.infoq.com/interviews/moore-apis-clocks
```
- <b> Observable </b>

```
  - Rather than calling a method, define a mechanism for retrieving and transforming the data, 
    - in the form of an “Observable,” 
  - and then subscribe an observer to it, 
  - the previously-defined mechanism fires into action 
  - observer captures and responds to its emissions whenever they are ready.

  - reference - http://reactivex.io/documentation/observable.html
```
- <b> Examples of Observable operators </b>

```
  - creation
  - transforming
  - filtering
  - combining
  - error handling
  - utility
  - conditional
  - mathematical
  - backpressure
  - connectable
  - converting

  - operators need to be chained to be useful.
```
- <b>FBP - Flow Based Programming</b>

```
  - defines apps as networks of "black box" processes
  - apps exchange data across predefined connections by message passing
  - connections are specified externally to the processes
  - the black box processes can be reconnected endlessly to form different apps 
    - no internal changes

  - reference: http://www.jpaulmorrison.com/fbp/
```
- <b>Actor Model</b>

```
  - actors are lightweight objects that communicate with each other
  - send receive messages
  - each actor has a mailbox
  - incoming messages are queued in the mailbox
  - actors have different states:
    - started, resumed, stopped & restarted
  - actors have an actor reference
    - reference does not change in case of actors restart etc
```
- <b>Reactive Coding</b>

```
  - When we query in a synchronous manner (which is what JDBC does) and 
    - wait for the result to come back from the database. 
  - During all of this time you will be monopolizing a thread 
    - that waits for the result to come back. 
  - A Java thread that just exists (without doing anything at all) 
    - can take up to 1 MB of heap memory, 
  - it is in your best interest to have quite a bit of memory available for your application 

  - <b>play framework</b> 
    - follows evented model than 1 user 1 thread model
  - each request is treated as an event
  - accessing db is one such event
  - run the events through a event loop
  - this results in not more threads created in the app [TEST]
  - play creates 1 thread per CPU core with a MAX 24 threads in the pool
  - still can do MORE than threaded model. Why ?
    - hand over the thread to other events when a task needs to do some waiting

  - reference - http://blog.jooq.org/2015/12/03/reactive-database-access-part-1-why-async/
```
- <b>Repeatable Read</b>

```
  - while in transaction, the first read operation to the
  - DB creates a snapshot of the entire DB so you are guarantee that all the
    - data in the DB will remain the same until the end of the transaction.

  - example:
    - ConnectionA selects from tableA in a transaction.
    - ConnectionB deletes all rows from tableB in a transaction.
    - ConnectionB commits.
    - ConnectionA loops over the rows of tableA and fetches from tableB using the 
      tableA_tableB_FK 
    - ConnectionA will get rows from tableB.
```
- <b>Read Committed</b>

```
  - while in a transaction, every query to the DB will get the committed data.

  - example:
    - ConnectionA starts a transaction.
    - ConnectionB starts a transaction.
    - ConnectionA insert row to tableA and commits.
    - ConnectionB insert row to tableA.
    - ConnectionB selects tableA and gets two rows.
    - ConnectionB commits / rollback.
```
- <b>Monitoring Specs </b>

```
  - need to feel the pulse
  - in other words health of the system
  - what is going on now
  - what has happened
  - status tracking
```
- <b>Events should be </b>

```
  - emitted by apps
  - available in seconds
  - not be lost often
  - cheap & scalable
```
- <b>Aggregations techniques</b>

```
  - heatmap
  - top N
  - histogram
  - count(), avg(), sum(), percent(), geo
```
- <b>References</b>

```
  - http://www.infoq.com/presentations/uber-stream-processing
  - https://github.com/openstack/mistral/commit/
    - 45493c2b0b33cabac60ab61b102552a70f295bb7
  - https://www.percona.com/blog/2015/01/14/
    - mysql-performance-implications-of-innodb-isolation-modes
  - https://www.percona.com/blog/2012/08/28/
    - differences-between-read-committed-and-repeatable-read-transaction-isolation-levels
```
